R/data-combined.R

```diff
@@ -0,0 +1,867 @@
+      } else {
+        dataCurrent <- dataNew
+      }
+
+      return(dataCurrent)
+    },
+
+    # transform the dataset using specified offsets and scale factors
+
+    .dataTransform = function(data,
+                              names = NULL,
+                              xOffsets = 0,
+                              yOffsets = 0,
+                              xScaleFactors = 1,
+                              yScaleFactors = 1) {
+
+      # return early if there is no data
+      if (is.null(data)) {
+        return(NULL)
+      }
+
+      # Keep all transformation parameters and their names linked together with
+      # dataframe data structure
+      #
+      # In case a list of parameters is specified, `flattenList()` will flatten
+      # it to an atomic vector of double type
+      #
+      # Additionally, if no names are provides, the transformations will apply
+      # to the entire dataframe, and thus dataset names can be placeholder just
+      # for the purpose of joining two dataframes
+      dataArg <- dplyr::tibble(
+        name          = names %||% unique(data$name),
+        xOffsets      = xOffsets,
+        yOffsets      = yOffsets,
+        xScaleFactors = xScaleFactors,
+        yScaleFactors = yScaleFactors
+      )
+
+      # update dataframe with transformation parameters
+      private$.dataTransformations <- private$.updateDF(private$.dataTransformations, dataArg)
+
+      # if present, remove old parameter columns
+      data <- dplyr::select(data, -dplyr::ends_with(c("Offsets", "ScaleFactors")))
+
+      # merge data with corresponding parameters and then
+      # replace `NA`s (which will be present for datasets for which no
+      # transformations were specified) with default values for offsets (0) and
+      # scale factors (1)
+      # these defaults mean no change is made to the data
+      data <- dplyr::left_join(data, private$.dataTransformations, by = "name") %>%
+        dplyr::mutate(across(matches("offsets$"), ~ tidyr::replace_na(.x, 0))) %>%
+        dplyr::mutate(across(matches("scalefactors$"), ~ tidyr::replace_na(.x, 1)))
+
+      # apply transformations
+      data <- dplyr::mutate(data,
+        xValues = (xOriginalValues + xOffsets) * xScaleFactors,
+        yValues = (yOriginalValues + yOffsets) * yScaleFactors,
+        yErrorValues = yOriginalErrorValues * yScaleFactors
+      )
+
+      return(data)
+    },
+
+    # add a new column with alternate names
+    #
+    # first group the dataset by its unique identifier, which is name, and then
+    # nest the rest of the dataframe in a list column, and then
+    # ungroup the dataframe as mutating column is not going to be by group, and then
+    # assign new `names` vector to the existing `name` column, and then
+    # unnest the list column
+
+    .renameDatasets = function(data, names = NULL) {
+      if (!is.null(names)) {
+        data <- data %>%
+          dplyr::group_by(name) %>%
```

Comment:
I think I need to understand the structure of the data before commenting any further. I am going to check unit tests